### PR TITLE
Make add_msg_debug and variants a macro.

### PR DESCRIPTION
**Pull request by prharvey**
#### Summary
None

#### Purpose of change

Parameters to `add_msg_debug` can be expensive, notably monster::name which shows up on profiles as taking up a few % of execution time.

#### Describe the solution

Turn `add_msg_debug` into a macro, which allows parameters to be evaluated only if the debug flag is actually enabled.

#### Testing

Verified that debug messages were still working as intended.

### CODE REVIEW

File: tools/clang-tidy-plugin/test/translations-in-debug-messages.cpp [hunk 1]
```diff
@@ -8,21 +8,6 @@
...
-
-    add_msg_debug( debugmode::DF_ACT_BUTCHER, pgettext( "Skill", "foo" ) );
-    // CHECK-MESSAGES: [[@LINE-1]]:47: warning: string arguments to debug message functions should not be translated, because this is an unnecessary performance cost. [cata-translations-in-debug-messages]
-
-    add_msg_debug( debugmode::DF_ACT_BUTCHER, n_gettext( "Skill", "Skills", 0 ) );
-    // CHECK-MESSAGES: [[@LINE-1]]:47: warning: string arguments to debug message functions should not be translated, because this is an unnecessary performance cost. [cata-translations-in-debug-messages]
-
-    add_msg_debug( debugmode::DF_ACT_BUTCHER, "%s", to_translation( "Skill" ) );
-    // CHECK-MESSAGES: [[@LINE-1]]:53: warning: string arguments to debug message functions should not be translated, because this is an unnecessary performance cost. [cata-translations-in-debug-messages]
-
```
Comment (Qrox) [inline tools/clang-tidy-plugin/test/translations-in-debug-messages.cpp:25]: Even if the performance cost is reduced, developer-facing debug messages should always be in English so these checks should be kept intact.

**Comment by prharvey** [comment]
Well that rebase when horribly wrong. Ignore this for now.

**Comment by prharvey** [comment]
Alright I fixed it, sorry for those who were notified.

**Comment by kevingranade** [comment]
Since this is perf,  can we see the before/after profile?

**Comment by prharvey** [comment]
> Since this is perf, can we see the before/after profile?

Since this has small effects everywhere it's hard to demonstrate directly, but I can show it indirectly by looking at monster::name

Before:
name

After:
Screenshot 2023-11-23 183230

Although it isn't the few % compared to master mentioned in the summary, as I first noticed this problem on my local version which has #69574 and #69198 on it (well, not the second one directly, since I'm rewriting that, but you get the idea), and those problems are dominating perf at the moment. Once they are in it is closer to 2% on just this function.

**Comment by akrieger** [comment]
If trying to demonstrate small-but-broad improvement, a good way to show it is to take several traces of a consistent time period (eg. wait the same 6 hours at a location) before & after to show overall runtime improvement. Note that I've found the profiler to have consistent variation of +-5% so improvements less than that may be lost to noise.

**Comment by prharvey** [comment]
> If trying to demonstrate small-but-broad improvement, a good way to show it is to take several traces of a consistent time period (eg. wait the same 6 hours at a location) before & after to show overall runtime improvement. Note that I've found the profiler to have consistent variation of +-5% so improvements less than that may be lost to noise.

This is actually why I mentioned it was hard to show directly currently. Against current master the performance is dominated by much larger problems so it is hard for this to stand out; right now the debug messages not being in a define is more of a code smell than a true performance issue.

That said, when the other changes I plan to make actually land, this will make a noticeable difference. Or at least a more convincing difference when you can actually see functions disappear from the flame graph. If there is any skepticism about the improvement, I've fine with waiting to merge this until those other things land as well.

I was originally going to get to this optimization much later anyways, but it was pretty easy to do independently and I was reminded of it by a recent NPC AI PR that was adding these extensively.

**Comment by akrieger** [comment]
Ah, ok. Well code looks fine to me and I understand the reasoning for it.

**Comment by akrieger** [comment]
I think it'll also seriously matter for #69623, I'm going to patch that and wait at refugee center and see how much it hurts.

**Comment by kevingranade** [comment]
```If there is any skepticism about the improvement, I'm fine with waiting to merge this until those other things land as well.```
I'm very much in the wait and see camp here,  yes it's potentially needless overhead,  but also it's overhead that is going to happen anyway if we are running with debug messages on,  and macros are gross. (This is not a criticism of you,  I'm not aware of a better way to make this happen either)

**Comment by prharvey** [comment]
> `If there is any skepticism about the improvement, I'm fine with waiting to merge this until those other things land as well.` I'm very much in the wait and see camp here, yes it's potentially needless overhead, but also it's overhead that is going to happen anyway if we are running with debug messages on, and macros are gross. (This is not a criticism of you, I'm not aware of a better way to make this happen either)

While I can't solve the issues with macros, the majority of users will be playing the game with debug messages off. So I think it is still important to remove overhead in this case.

**Pull request merged**